src/formatting_stack/reporters/pretty_line_printer.clj

```diff
@@ -0,0 +1,79 @@
+(ns formatting-stack.reporters.pretty-line-printer
+  "Outputs report similar to github.com/xojs/xo"
+  (:require
+   [clojure.stacktrace :refer [print-stack-trace]]
+   [formatting-stack.protocols.reporter :as reporter]
+   [formatting-stack.protocols.spec :as protocols.spec]
+   [formatting-stack.reporters.impl :refer [truncate-line-wise]]
+   [formatting-stack.util :refer [colorize]]
+   [medley.core :refer [map-vals]]
+   [nedap.speced.def :as speced]
+   [nedap.utils.modular.api :refer [implement]]))
+
+(speced/defn print-summary [{:keys [^boolean? summary?
+                                    ^boolean? colorize?]} reports]
+  (when summary?
+    (->> reports
+         (group-by :level)
+         (map-vals count)
+         (into (sorted-map-by compare)) ;; print summary in order
+         (run! (fn [[report-type n]]
+                 (cond-> (str n (case report-type
+                                  :exception " exceptions occurred"
+                                  :error     " errors found"
+                                  :warning   " warnings found"))
+                   colorize? (colorize (case report-type
+                                         :exception :red
+                                         :error     :red
+                                         :warning   :yellow))
+                   true      println))))))
+
+(speced/defn print-exceptions [{:keys [^boolean? print-stacktraces?]} reports]
+  (->> reports
+       (filter (speced/fn [{:keys [^::protocols.spec/level level]}]
+                 (#{:exception} level)))
+       (group-by :filename)
+       (run! (fn [[title reports]]
+               (println (colorize title :cyan))
+               (doseq [{:keys [^Throwable exception]} reports]
+                 (if print-stacktraces?
+                   (print-stack-trace exception)
+                   (println (ex-message exception)))
+                 (println))))))
+
+(speced/defn print-warnings [{:keys [max-msg-length
+                                     ^boolean? colorize?]}
+                             ^::protocols.spec/reports reports]
+  (->> (filter (fn [{:keys [level]}] (#{:error :warning} level)) reports)
+       (group-by :filename)
+       (into (sorted-map-by compare))
+       (run! (fn [[title reports]]
+               (println (cond-> title colorize? (colorize :cyan)))
+               (doseq [{:keys [msg msg-extra-data column line source level]} (sort-by :line reports)]
+                 (println (case level
+                            :error   " ❌"
+                            :warning " ⚠️")
+                          (if (or column line)
+                            (cond-> (format "%3s:%-3s" (or line "?") (or column "?")) colorize? (colorize :grey))
+                            "       ")
```

Comment:
Perhaps a bit overly brancing, would personally not mind `?:?` as a line:col number.